### PR TITLE
Turing - Advanced Classes Implementation (Audited Pointer and AnnotatedWrapper)

### PR DESCRIPTION
**Pull request by caden-james**
This pull request contains our updated functionality for our AuditedPointer and AnnotatedWrapper classes. Desired utility is outlined in the Class Specs folder.

### CODE REVIEW

File: Group-08/Dev/AuditedPointer/AuditedPointer.hpp [hunk 1]
```diff
@@ -78,42 +79,86 @@ Aptr<TYPE> MakeAudited(Args&&... args)
...
+    }
+
+    // Get the active Aptr map
+    std::unordered_map<int, const void*> GetActiveAptrs() const
+    {
+        return activeAptrs;
+    }
+
+    // Find an Aptr by ID
+    const void* FindAptr(int id) const
```
Comment (ShahaabAli) [inline Group-08/Dev/AuditedPointer/AuditedPointer.hpp:123]: Could return std::optional<const void*> instead to avoid returning a nullptr enable the following logic:

auto maybePtr = leakChecker.FindAptr(someID);
if (maybePtr) {
    std::cout << "Found pointer: " << *maybePtr;
} else {
    std::cout << "Pointer not found";
}

would return std::nullopt instead of nullptr

File: Group-08/Dev/AuditedPointer/AuditedPointer.hpp
Comment (ShahaabAli) [inline Group-08/Dev/AuditedPointer/AuditedPointer.hpp:1]: Overall, this class is implemented very well. It is clear and follows the C++ standard. I would recommend adding doxygen comments, especially at the beginning of structs and classes to clearly express their purpose. Additionally, comments in the function bodies could add more clarity to what exactly is being done. Although you are lacking std algorithms, constexpr, and lambda functions, I am unsure where they can be used within your class to improve it. Aside from that, I added some small comments, but overall, good job!

File: Group-08/Dev/Tests/AdvancedTests/AnnotatedWrapperAdvancedTest.cpp
Comment (ShahaabAli) [inline Group-08/Dev/Tests/AdvancedTests/AnnotatedWrapperAdvancedTest.cpp:1]: These tests seems to comprehensively test Annotated Wrapper. I have a few small tests that may be worth adding. Testing clearAnnotation after use. Testing listAnnotation on empty annotation. Testing on types outside of strings and numeric types may be useful to test the templating capability, but this depends on your planned application of this class.

File: Group-08/Dev/Tests/AdvancedTests/AuditedPointerAdvancedTest.cpp
Comment (ShahaabAli) [inline Group-08/Dev/Tests/AdvancedTests/AuditedPointerAdvancedTest.cpp:1]: Overall, this is pretty good testing as well. Here are some small tests that may be worth adding. Test move constructor, assignment operator, and find function

Comment (asifmasood7) [inline Group-08/Dev/Tests/AdvancedTests/AuditedPointerAdvancedTest.cpp:1]: After reviewing this part, I believe that your tests do cover the basic functionality, such as verifying unique ID assignment and tracking active pointers. Especially the checks for deletion and the reset functionality are clear and effective.

I think it would be beneficial to add some edge-case tests as I mentioned in .hpp review as well. For example, tests that attempt to dereference or delete a null pointer would help ensure that your error handling is robust. I wanted to emphasize on that because I think that will align well with our emphasis on thorough error checking.

Overall, the tests are well-organized and easy to understand. They seem to effectively confirm that active pointers are tracked correctly and that the reset function clears the active set. Awesome work!

**Review by asifmasood7 — COMMENTED** [review]
I have finished the reviews for Group 8 AuditedPointer and AnnotatedWrapper Advanced version classes by trying to focus on assigned part for "Testing & Error Checking & Code's Optimization and Flexibility." Thank you!

File: Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+        // Iterates through all annotations and applies a callback
+        void listAnnotations(std::function<void(const std::string&, const T&)> callback) const {
+            for (const auto& [key, value] : mAnnotations) {
+                callback(key, value);
+            }
+        }
+
+        // Sets the font size 
+        void setFontSize(int size) {
```
Comment (MaxKrawec) [inline Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp:115]: Intuitive and functional interface: Is it a good idea to add a default argument for the size parameter here?

File: Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+            mFontSize = size;
+        }
+
+        // Retrieves the current font size
+        int getFontSize() const {
+            return mFontSize;
+        }
+
+        // Sets the font color
+        void setFontColor(int color) {
```
Comment (MaxKrawec) [inline Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp:125]: Intuitive and functional interface: Is it a good idea to add a default argument for the color parameter here? Maybe a black or white color value?

File: Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+            mFontColor = color;
+        }
+
+        // Retrieves the current font color
+        int getFontColor() const {
+            return mFontColor;
+        }
+
+        // Sets the background color
+        void setBackgroundColor(int color) {
```
Comment (MaxKrawec) [inline Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp:135]: Intuitive and functional interface: Is it a good idea to add a default argument for the color parameter here? Maybe a black or white color value?

File: Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+namespace cse {
+
+    // A template class that stores annotations and visual styling.
+    template<typename T>
+    class AnnotatedWrapper {
+    private:
+        // A map to hold annotation key-value pairs.
+        std::unordered_map<std::string, T> mAnnotations;
+
+        // Visual styling properties
```
Comment (MaxKrawec) [inline Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp:79]: Informative comments: Good job adding comments to the member variables.

File: Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        int mFontSize = 0;
+        int mFontColor = 0;
+        int mBackgroundColor = 0;
+
+    public:
```
Comment (MaxKrawec) [inline Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp:84]: Informative comments: Few notes:
1) The member function comment helps someone new to the code understand it more rapidly.
2) The class avoids redundant/obvious comments that clutter the code.

Potential Suggestion:
Personally, I suggest using doxygen style comments for functions. It is a good, consistent template for clearly explaining what a function does. I especially like it for the @param and @return comments, so I clearly know what the input and output of a function should be.

File: Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp [hunk 7]
```diff
@@ -0,0 +1,147 @@
+#ifndef ANNOTATEDWRAPPER_H
+#define ANNOTATEDWRAPPER_H
```
Comment (MaxKrawec) [inline Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp:2]: Consistent style: It looks like in some files you use `#pragma once`, and in others, you use this method. Either option is fine, but I would pick one to be consistent.

File: Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+    template<typename T>
+    class AnnotatedWrapper {
+    private:
+        // A map to hold annotation key-value pairs.
+        std::unordered_map<std::string, T> mAnnotations;
+
+        // Visual styling properties
+        int mFontSize = 0;
+        int mFontColor = 0;
+        int mBackgroundColor = 0;
```
Comment (MaxKrawec) [inline Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp:82]: Consistent style: Good job using the m prefix for member variables

File: Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp (hunk 6 above)
Comment (MaxKrawec) [reply Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp:84]: Consistent style: Additionally, it looks like in other files you use doxygen style comments. To be consistent, I would add them here.

**Review by MaxKrawec — COMMENTED** [review]
Great job overall! My comments are inline in the file.

File: Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+=======
+#include <unordered_map>
+#include <string>
+#include <iostream>
+#include <functional>
+
+namespace cse {
+
+    // A template class that stores annotations and visual styling.
+    template<typename T>
```
Comment (MaxKrawec) [inline Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp:73]: Informative identifiers: Potentially change the template parameter (T) to be more descriptive.

**Review by MaxKrawec — COMMENTED** [review]
Good job overall! The review comments are inline in the files.

File: Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp [hunk 10]
```diff
@@ -0,0 +1,147 @@
+#ifndef ANNOTATEDWRAPPER_H
```
Comment (PMitkiewicz) [inline Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp:1]: Overall impressions on Annotated Wrapper. Not much to critique, this is a very well designed class with a very logical structure. There are a few places that can be optimized, such as possibly using set size arrays to improve cleanup, and using std::optional for getAnnotation. Furthermore, removeAnnotations can check if the key exists before deleting it to slightly improve performance. In terms of structure, the only thing I would say is you should maybe split up the class into a helper class. If we think of it in terms of the purpose of the class, it's really doing two things, the annotations and the prettifying and visual stuff. Perhaps putting the visual effects in a separate class would be more consistent. Other than that, you did a very good job of keeping the class efficient, optimized, and with the correct modularity. Well done!

File: Group-08/Dev/AuditedPointer/AuditedPointer.hpp [hunk 11]
```diff
@@ -78,42 +79,86 @@ Aptr<TYPE> MakeAudited(Args&&... args)
...
+    }
+
+    // Remove an Aptr from the active set
+    void RemoveAptr(int id)
+    {
+        activeAptrs.erase(id);
+    }
+
+    // Get the active Aptr map
+    std::unordered_map<int, const void*> GetActiveAptrs() const
```
Comment (PMitkiewicz) [inline Group-08/Dev/AuditedPointer/AuditedPointer.hpp:117]: Again, is it possible to use a less expensive container than an unordered map?

File: Group-08/Dev/AuditedPointer/AuditedPointer.hpp (hunk 1 above)
Comment (PMitkiewicz) [reply Group-08/Dev/AuditedPointer/AuditedPointer.hpp:123]: I agree, std::optional would increase performance

File: Group-08/Dev/AuditedPointer/AuditedPointer.hpp [hunk 12]
```diff
@@ -124,22 +169,23 @@ class Aptr
...
+                leakChecker.AddAptr(mID, mPtr);
             }
         }
         return *this;
     }
 
     // Move constructor
-    Aptr(Aptr&& other) : mPtr(other.mPtr)
+    Aptr(Aptr&& other) : mPtr(other.mPtr), mID(other.mID)
     {
```
Comment (PMitkiewicz) [inline Group-08/Dev/AuditedPointer/AuditedPointer.hpp:186]: I would break up this constructor. It is currently quite large, and sort of difficult to follow. I would recommend creating templated helper functions that you can call on multiple constructors, as I believe there is some overlap in functionality. If that's not possible, just make individual helper functions, or add better comments.

File: Group-08/Dev/AuditedPointer/AuditedPointer.hpp [hunk 13]
```diff
@@ -163,7 +211,7 @@ class Aptr
...
         if (!mPtr)
         {
-            std::cerr << "Dereferencing a null pointer" << std::endl;
+            std::cerr << "Dereferencing a null pointer (ID: " << mID << ")" << std::endl;
             std::terminate();
         }
         return *mPtr;
```
[the remark this answers is not in the document]
Comment (PMitkiewicz) [reply Group-08/Dev/AuditedPointer/AuditedPointer.hpp:217]: I agree, either that or break it up, function is hard to follow currently.

File: Group-08/Dev/AuditedPointer/AuditedPointer.hpp [hunk 14]
```diff
@@ -191,54 +239,61 @@ class Aptr
...
     // Get a set of all active AuditedPointer IDs and their pointers
-    static std::unordered_map<int, void*> GetActiveAptrs()
+    static std::unordered_map<int, const void*> GetActiveAptrs()
     {
-        return {};
+        return leakChecker.GetActiveAptrs();
     }
     
-        // Find an AuditedPointer by ID in the active set
-    static void* Find(int id)
```
Comment (PMitkiewicz) [inline Group-08/Dev/AuditedPointer/AuditedPointer.hpp:215]: Good use of advanced C++ features

File: Group-08/Dev/AuditedPointer/AuditedPointer.hpp [hunk 15]
```diff
@@ -191,54 +239,61 @@ class Aptr
...
-    ++globalCounter;
+    TYPE* ptr = new TYPE(std::forward<Args>(args)...);
+    return Aptr<TYPE>(ptr);
 }
 
 #endif
 
-}
\ No newline at end of file
+}
```
Comment (PMitkiewicz) [inline Group-08/Dev/AuditedPointer/AuditedPointer.hpp:299]: My overall review on the audited pointer class. Fantastic job, your use of static functions and templates is well chosen. Comments are informative, though I believe there could be more extensive explanation on the complex functions, as I commented on. Overall, it is well structured but I believe it can be improved. Look for places you can abstract, as I believe you can either find redundancies across multiple functions (like the constructors), or you have functions that do multiple things, that you should make into smaller helper functions. Furthermore, the use of std::optional can help performance in certain classes, and if possible, see if you can make the class work with a static size container rather than unordered map. Overall though, great job!

File: Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (ellio357) [reply Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp:82]: Will handle colors on the front-end of our application

File: Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+
+=======
+#include <unordered_map>
+#include <string>
+#include <iostream>
+#include <functional>
+
+namespace cse {
+
+    // A template class that stores annotations and visual styling.
```
[the remark this answers is not in the document]
Comment (ellio357) [reply Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp:72]: done

File: Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+        // Sets the background color
+        void setBackgroundColor(int color) {
+            mBackgroundColor = color;
+        }
+
+        // Retrieves the current background color
+        int getBackgroundColor() const {
+            return mBackgroundColor;
+        }
+>>>>>>> 11b16d6c263a71a2ef9ac0b7df019ccc695a6518
```
[the remark this answers is not in the document]
Comment (ellio357) [reply Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp:143]: resolved

File: Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+ *   annotations as key-value pairs.
+ *
+ */
+namespace cse {
+
+    class AnnotatedWrapper {
+    public:
+        // Constructor
+        AnnotatedWrapper(){
+            mAnnotations = std::unordered_map<std::string, std::string>();
```
[the remark this answers is not in the document]
Comment (ellio357) [reply Group-08/Dev/AnnotatedWrapper/AnnotatedWrapper.hpp]: great note!

File: Group-08/Dev/AuditedPointer/AuditedPointer.hpp
Comment (caden-james) [reply Group-08/Dev/AuditedPointer/AuditedPointer.hpp:1]: Totally understand what you're saying and appreciate the comments. I added in updated functionality included more constexpr functions when possible and a lambda function for the MakeAudited function.

File: Group-08/Dev/AuditedPointer/AuditedPointer.hpp [hunk 19]
```diff
@@ -3,6 +3,7 @@
 #include <iostream>
```
[the remark this answers is not in the document]
Comment (caden-james) [reply Group-08/Dev/AuditedPointer/AuditedPointer.hpp:3]: Both files use `#pragma once` now.

File: Group-08/Dev/AuditedPointer/AuditedPointer.hpp [hunk 20]
```diff
@@ -78,42 +79,86 @@ Aptr<TYPE> MakeAudited(Args&&... args)
 
 // ==================== Debug Mode (Audited Pointer) ====================
 
-// Global counter and LeakChecker
-int globalCounter = 0; 
+template <typename TYPE>
+class Aptr;
 
+// LeakChecker struct to track active Aptr objects
 struct LeakChecker
```
[the remark this answers is not in the document]
Comment (caden-james) [reply Group-08/Dev/AuditedPointer/AuditedPointer.hpp:90]: I'm just going to keep it as LeakChecker as I think the meaning is obvious and I don't want variables to get too wordy to make future updates more difficult. The function seems implicit to me.

File: Group-08/Dev/AuditedPointer/AuditedPointer.hpp [hunk 21]
```diff
@@ -78,42 +79,86 @@ Aptr<TYPE> MakeAudited(Args&&... args)
...
+        activeAptrs.clear();
+    }
 };
 
-LeakChecker leakChecker; // Global instance of LeakChecker
+// Global instance of LeakChecker
+static LeakChecker leakChecker;
 
 template <typename TYPE>
 class Aptr
```
[the remark this answers is not in the document]
Comment (caden-james) [reply Group-08/Dev/AuditedPointer/AuditedPointer.hpp:142]: I feel like Aptr implies AuditedPointer considering the file name so it seems reasonable enough to me.

File: Group-08/Dev/AuditedPointer/AuditedPointer.hpp [hunk 22]
```diff
@@ -78,42 +79,86 @@ Aptr<TYPE> MakeAudited(Args&&... args)
...
+    {
+        activeAptrs.clear();
+    }
 };
 
-LeakChecker leakChecker; // Global instance of LeakChecker
+// Global instance of LeakChecker
+static LeakChecker leakChecker;
 
 template <typename TYPE>
```
[the remark this answers is not in the document]
Comment (caden-james) [reply Group-08/Dev/AuditedPointer/AuditedPointer.hpp:141]: Considering the type could be literally anything as the underlying pointer could be anything, I don't see how it could be more descriptive.

File: Group-08/Dev/AuditedPointer/AuditedPointer.hpp (hunk 11 above)
Comment (caden-james) [reply Group-08/Dev/AuditedPointer/AuditedPointer.hpp:117]: I figured since we want each underlying Aptr (void *) to be attached to an int (ID #), an unordered map does this implicity. I suppose a list could have similar functionality but both would have similar overheads aswell.

File: Group-08/Dev/AuditedPointer/AuditedPointer.hpp (hunk 1 above)
Comment (caden-james) [reply Group-08/Dev/AuditedPointer/AuditedPointer.hpp:122]: Didn't think of this, now implemented!

File: Group-08/Dev/AuditedPointer/AuditedPointer.hpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (caden-james) [reply Group-08/Dev/AuditedPointer/AuditedPointer.hpp:142]: Added the class comment.

File: Group-08/Dev/AuditedPointer/AuditedPointer.hpp (hunk 12 above)
Comment (caden-james) [reply Group-08/Dev/AuditedPointer/AuditedPointer.hpp:192]: Not sure exactly which one is being referenced here but if it's the copy constructor that one has been cleaned up and shortened greatly.

File: Group-08/Dev/AuditedPointer/main.cpp [hunk 23]
```diff
@@ -1,10 +1,10 @@
-#include "AuditedPointer.h"
+#include "AuditedPointer.hpp"
 #include <vector>
 
 int main()
 {
     // Example usage
-    Aptr<int> ptr = MakeAudited<int>(42);
+    cse::Aptr<int> ptr = cse::MakeAudited<int>(42);
     std::cout << *ptr << std::endl; // Output: 42
```
[the remark this answers is not in the document]
Comment (caden-james) [reply Group-08/Dev/AuditedPointer/main.cpp:8]: I am a bit confused, do you mean to compare it to the Aptr class? This was mainly just to show off the functionality of the Aptr class.

File: Group-08/Dev/AuditedPointer/AuditedPointer.hpp [hunk 24]
```diff
@@ -0,0 +1,426 @@
...
+    }
+
+    /**
+     * @brief Gets all currently active Aptrs
+     * @return Map of active pointer IDs to their addresses
+     */
+    static std::unordered_map<int, const void*> GetActiveAptrs()
+    {
+        return leakChecker.GetActiveAptrs();
+    }
```
Comment (mercere99) [inline Group-08/Dev/AuditedPointer/AuditedPointer.hpp:371]: This function should return a const reference, I think.  No need to copy the entire map...

File: Group-08/Dev/AuditedPointer/AuditedPointer.hpp (hunk 11 above)
Comment (mercere99) [reply Group-08/Dev/AuditedPointer/AuditedPointer.hpp:185]: That said, this function should return a const reference rather than copying the entire map (which is an unneeded expense)

**Review by mercere99 — COMMENTED** [review]
Great job with both `AuditedPointer` and `AnnotatedWrapper`

Some notes:
- Please be careful to make sure that the final version for review compiles correctly (in this case it had a merge conflict, though it was fixed speedily after being pointed out.)
  - The copy construct and assignment operator for the audited pointer should make a copy of the pointer itself, NOT a deep copy of the object being pointed to.  It's fine if you want to add new functionality for a deep copy, but the class show otherwise match the functionality of a raw pointer (more details inline).
  - I made lots of code suggestions for expanding Aptr.  I want to stress that I am very impressed by the class as is and that these suggestions are just for making it more well rounded.

Nicely done!
